library(tidyverse)

### figure 5A, 
# Tip 1: NetSurfP - 3.0 does not accept protein sequence more than 5,000 residues. 
# You can manually check the sequence length and split longer sequences into shorter segments.
# Tip 2: NetSurfP-3.0 does not accept U as an amino acid residue.
# You can use the GRAVY Calculator (https://www.gravy-calculator.de/index.php) to locate U residues in your FASTA file and replace them with C.


### figure 5C, ubiquitination sites occupancy




